File: languages/typescript/02-core/error-handling-and-defensive-programming/example/main.ts
// Module focus: Guarding risky inputs so failures stay explicit and controlled.
// Why it matters: practicing error handling and defensive programming patterns makes exercises and checkpoints easier to reason about.

type ScoreRecord = {
    name: string;
    score: number;
};

function parseRecord(line: string): ScoreRecord | null {
    // Treat the last field as the score so names can contain spaces.
    const parts = line.trim().split(/\s+/);
    if (parts.length < 2) {
        return null;
    }

    // Defensive parsing rejects non-numeric and out-of-range scores.
    const score = Number.parseInt(parts.at(-1) ?? "", 10);
    if (!Number.isInteger(score) || score < 0 || score > 100) {
        return null;
    }

    // A valid score is not enough; the record also needs a name.
    const name = parts.slice(0, -1).join(" ");
    if (!name) {
        return null;
    }

    return { name, score };
}

const rows = [
    "Ana Smith 91",
    "InvalidRow",
    "Bob Lee ninety",
    "Carla Mendez 88",
];
const accepted: ScoreRecord[] = [];

// Process every row independently so one bad value does not stop the whole batch.
for (const row of rows) {
    const record = parseRecord(row);
    if (record === null) {
        // Report output values so learners can verify the error handling and defensive programming result.
        console.log(`Skipped invalid row: ${row}`);
        continue;
    }
    accepted.push(record);
}

// Report only records that survived every validation check.
console.log(`Accepted rows: ${accepted.length}`);
for (const record of accepted) {
    console.log(`- ${record.name}: ${record.score}`);
}

export {};

File: languages/java/02-core/error-handling-and-defensive-programming/example/Main.java
// Module focus: Guarding risky inputs so failures stay explicit and controlled.
// Why it matters: practicing error handling and defensive programming patterns makes exercises and checkpoints easier to reason about.

public class Main {
    static int divide(int left, int right) {
        if (right == 0) {
            // Guard before the risky operation so the failure has a clear explanation.
            throw new ArithmeticException("division by zero");
        }
        return left / right;
    }

    public static void main(String[] args) {
        String[] rows = {"42 6", "18 0", "bad row"};

        for (String row : rows) {
            try {
                String[] parts = row.split("\\s+");
                // Parsing is intentionally inside the try block because malformed rows are expected.
                int left = Integer.parseInt(parts[0]);
                int right = Integer.parseInt(parts[1]);

                // Report valid calculations and recoverable failures from the same loop.
                System.out.println("Input: " + left + " " + right);
                System.out.println("Result: " + divide(left, right));
            } catch (ArithmeticException error) {
                System.out.println("Cannot divide by zero.");
            } catch (RuntimeException error) {
                // Other runtime failures are treated as malformed input rows for this exercise.
                System.out.println("Skipped invalid row: " + row);
            }
        }
    }
}
